### PR TITLE
log: allow setting `vmodule` on-the-fly

### PR DESCRIPTION
**Pull request by tbg**
Issue a GET request to /debug/vmodule/<vmodule_spec> to use this feature.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/8948)

### CODE REVIEW

**Comment by petermattis** [comment]
:lgtm:

---

Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[server/debug.go, line 126 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8948#-KQRAQWBZyoDfM05tkpF:-KQRAQWBZyoDfM05tkpG:bzf490h) ([raw file](https://github.com/cockroachdb/cockroach/blob/64b516f0d225d84d7f5aad3622e3b5c1e14c1a27/server/debug.go#L126)):*
> ```Go
> <td>change vmodule</td>
> <td>
> get /debug/setvmodule/<your_vmodule_here><br />For example, <code>*=1</code> or <code>raft=3,storage=2</code>. Empty string disables vmodule logging.
> ```

`s/setvmodule/vmodule/g`?

---

*[util/log/log.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8948#-KQRAWMIajRvqbIrbtTp:-KQRAWMIajRvqbIrbtTq:bj4cmd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/64b516f0d225d84d7f5aad3622e3b5c1e14c1a27/util/log/log.go#L31)):*
> ```Go
> 	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
> 	spec := r.RequestURI[len(httpLogLevelPrefix):]
> 	if err := logging.vmodule.Set(spec); err != nil {
> ```

I would think the vmodule value should be a query parameter, though perhaps that is awkward given the value contains `=`.

---

*[util/log/log.go, line 39 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8948#-KQRAj49YU_QWcriF9u4:-KQRAj49YU_QWcriF9u5:bpktfsf) ([raw file](https://github.com/cockroachdb/cockroach/blob/64b516f0d225d84d7f5aad3622e3b5c1e14c1a27/util/log/log.go#L39)):*
> ```Go
> 
> func init() {
> 	http.Handle(httpLogLevelPrefix, http.HandlerFunc(handleVModule))
> ```

There is another `init` just a few lines below. Add this there? Or perhaps the `log` package should expose the mechanism to change the vmodule but leave it to some other package to add the http hook.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8948#-:-KQRBHW-jeGYv21mHEO3:bnfp4nl)*

**Comment by tschottdorf** [comment]
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[server/debug.go, line 126 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8948#-KQRAQWBZyoDfM05tkpF:-KQRBXE94dIYJ0nQnCoz:bgds2dm) ([raw file](https://github.com/cockroachdb/cockroach/blob/64b516f0d225d84d7f5aad3622e3b5c1e14c1a27/server/debug.go#L126)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/setvmodule/vmodule/g`?
</details>
No rename without an omission. Done.

---

*[util/log/log.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8948#-KQRAWMIajRvqbIrbtTp:-KQRBqQ5YkLo60U4r_dk:bvxf80t) ([raw file](https://github.com/cockroachdb/cockroach/blob/64b516f0d225d84d7f5aad3622e3b5c1e14c1a27/util/log/log.go#L31)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I would think the vmodule value should be a query parameter, though perhaps that is awkward given the value contains `=`.
</details>
I'll leave as is for now.

---

*[util/log/log.go, line 39 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8948#-KQRAj49YU_QWcriF9u4:-KQRBlTQl3QjNTrygmmA:bd5ebkh) ([raw file](https://github.com/cockroachdb/cockroach/blob/64b516f0d225d84d7f5aad3622e3b5c1e14c1a27/util/log/log.go#L39)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> There is another `init` just a few lines below. Add this there? Or perhaps the `log` package should expose the mechanism to change the vmodule but leave it to some other package to add the http hook.
</details>
Merged them. I'm not interested in getting fancy here at this point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8948)*

File: util/log/log.go [hunk 1]
```diff
@@ -16,9 +16,27 @@
...
+
+const httpLogLevelPrefix = "/debug/vmodule/"
+
+func handleVModule(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+	spec := r.RequestURI[len(httpLogLevelPrefix):]
+	if err := logging.vmodule.Set(spec); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+	}
+	Infof(context.Background(), "vmodule changed to: %s", spec)
```
Comment (tamird) [inline util/log/log.go]: for the future: you can use `r.Context()` here.

**Comment by bdarnell** [comment]
We'll need to put this behind authentication (or disable it for release builds) before 1.0. Otherwise it's an easy DoS vector (turning on maximum verbosity is really expensive). 

---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[util/log/log.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8948#-KQRAWMIajRvqbIrbtTp:-KQT4P_WlIsnI-BJ10sc:b-am41u) ([raw file](https://github.com/cockroachdb/cockroach/blob/64b516f0d225d84d7f5aad3622e3b5c1e14c1a27/util/log/log.go#L31)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I'll leave as is for now.
</details>
A query parameter would make it trivial to make a simple html form on /debug/, and then the browser would encode the `=` signs for you (if that's even necessary - i think equals signs are only *required* to be escaped in keys, not values). Or you could do something like `/vmodule?raft=1&store=2`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8948#-:-KQT59bjXbKG65h6emC3:b-c1elpx)*

**Comment by tschottdorf** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[util/log/log.go, line 31 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8948#-KQRAWMIajRvqbIrbtTp:-KQT6nqtxMynucY-pW0-:btz5k1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/64b516f0d225d84d7f5aad3622e3b5c1e14c1a27/util/log/log.go#L31)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> A query parameter would make it trivial to make a simple html form on /debug/, and then the browser would encode the `=` signs for you (if that's even necessary - i think equals signs are only *required* to be escaped in keys, not values). Or you could do something like `/vmodule?raft=1&store=2`.
</details>
Not opposed to it, just don't think it's worth the (my) time at this point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8948)*

**Pull request merged**